File: src/activation.cpp
#include "activation.hpp"

#include <algorithm>
#include <cmath>

namespace volt {

float apply_activation(float I, Activation a, const Config& cfg) {
    switch (a) {
        case Activation::Identity:
            return I;
        case Activation::ReLU:
            return std::max(0.0f, I);
        case Activation::Sigmoid: {
            const float mid = cfg.I_min + 0.5f * cfg.I_range;
            const float scale = cfg.I_range * 0.25f + 1e-30f;
            const float x = (I - mid) / scale * cfg.activation_sigmoid_steepness;
            const float s = 1.0f / (1.0f + std::exp(-x));
            return cfg.I_min + cfg.I_range * s;
        }
        default:
            return I;
    }
}

double apply_activation(double I, Activation a, const Config& cfg) {
    switch (a) {
        case Activation::Identity:
            return I;
        case Activation::ReLU:
            return std::max(0.0, I);
        case Activation::Sigmoid: {
            const double mid = static_cast<double>(cfg.I_min) + 0.5 * static_cast<double>(cfg.I_range);
            const double scale = static_cast<double>(cfg.I_range) * 0.25 + 1e-30;
            const double x =
                (I - mid) / scale * static_cast<double>(cfg.activation_sigmoid_steepness);
            const double s = 1.0 / (1.0 + std::exp(-x));
            return static_cast<double>(cfg.I_min) + static_cast<double>(cfg.I_range) * s;
        }
        default:
            return I;
    }
}

}  // namespace volt

File: crates/volt-core/src/activation.rs
use crate::config::Config;

#[derive(Debug, Clone, Copy, PartialEq, Default)]
pub enum Activation {
    #[default]
    Identity,
    ReLU,
    Sigmoid,
}

pub fn apply_activation_f32(i: f32, a: Activation, cfg: &Config) -> f32 {
    match a {
        Activation::Identity => i,
        Activation::ReLU => i.max(0.0),
        Activation::Sigmoid => {
            let mid = cfg.i_min + 0.5 * cfg.i_range;
            let scale = cfg.i_range * 0.25 + 1e-30;
            let x = (i - mid) / scale * cfg.activation_sigmoid_steepness;
            let s = 1.0 / (1.0 + (-x).exp());
            cfg.i_min + cfg.i_range * s
        }
    }
}

pub fn apply_activation_f64(i: f64, a: Activation, cfg: &Config) -> f64 {
    match a {
        Activation::Identity => i,
        Activation::ReLU => i.max(0.0),
        Activation::Sigmoid => {
            let mid = cfg.i_min as f64 + 0.5 * cfg.i_range as f64;
            let scale = cfg.i_range as f64 * 0.25 + 1e-30;
            let x = (i - mid) / scale * cfg.activation_sigmoid_steepness as f64;
            let s = 1.0 / (1.0 + (-x).exp());
            cfg.i_min as f64 + cfg.i_range as f64 * s
        }
    }
}
